src/context/Discussions/DiscussionProvider.tsx

```diff
@@ -0,0 +1,167 @@
+  handleDeleteComment: (_ID, _msgID) => {},
+  permissions: {
+    canPost: false,
+  },
+  loading: false,
+  posting: false,
+  deleting: false,
+});
+
+interface DiscussionProviderProps {}
+
+const sortMessages = (messages: MessageDetailsFragment[] = []) => sortBy(messages, item => item.timestamp);
+
+const DiscussionProvider: FC<DiscussionProviderProps> = ({ children }) => {
+  const handleError = useApolloErrorHandler();
+  const { discussionId } = useUrlParams();
+  const { ecoverseNameId, loading: loadingEcoverse } = useEcoverse();
+  const { communityId, loading: loadingCommunity } = useCommunityContext();
+
+  const { data, loading } = useCommunityDiscussionQuery({
+    variables: { ecoverseId: ecoverseNameId, communityId, discussionId },
+    skip: !communityId || !ecoverseNameId || !discussionId,
+  });
+
+  const discussionData = data?.ecoverse.community?.communication?.discussion;
+
+  const senders = uniq(discussionData?.messages?.map(m => m.sender) || []);
+  const { getAuthor, authors, loading: loadingAuthors } = useAuthorsDetails(senders);
+
+  const sortedMessages = sortMessages(discussionData?.messages || []);
+
+  const discussion =
+    discussionData &&
+    ({
+      id: discussionData.id,
+      title: discussionData.title,
+      category: discussionData.category,
+      myPrivileges: discussionData.authorization?.myPrivileges ?? [],
+      author: getAuthor(discussionData.createdBy || ''),
+      authors: authors,
+      description: discussionData.description,
+      createdAt: discussionData.timestamp ? new Date(discussionData.timestamp) : new Date(),
+      totalComments: discussionData.commentsCount,
+      comments: sortedMessages.map<Comment>(m => ({
+        id: m.id,
+        body: m.message,
+        author: getAuthor(m.sender),
+        createdAt: new Date(m.timestamp),
+      })),
+    } as Discussion);
+
+  const [postComment, { loading: postingComment }] = usePostDiscussionCommentMutation();
+
+  const handlePostComment = async (discussionId: string, post: string) => {
+    await postComment({
+      update(cache, { data }) {
+        cache.modify({
+          id: cache.identify({
+            id: discussionId,
+            __typename: 'Discussion',
+          }),
+          fields: {
+            messages(existingMessages = []) {
+              if (data) {
+                const newMessage = cache.writeFragment({
+                  data: data?.sendMessageToDiscussion,
+                  fragment: MessageDetailsFragmentDoc,
+                });
+                return [...existingMessages, newMessage];
+              }
+              return existingMessages;
+            },
+          },
+        });
+      },
+      variables: {
+        input: {
+          discussionID: discussionId,
+          message: post,
+        },
+      },
+    });
+  };
+
+  const permissions: Permissions = useMemo(
+    () => ({
+      canPost: true, // TODO [ATS]: Should be populated.
```

Comment:
use `myPrivileges`?
What about deleting?
Right now permissions are decided in `DiscussionView`. And comment permissions are decided case by case in the same view. Is there a way to generalize?

Reply:
I'll remove it.